appbuilder/services/livesync/livesync-service.ts

```diff
@@ -0,0 +1,102 @@
+///<reference path="../../../.d.ts"/>
+"use strict";
+
+import { exportedPromise } from "../../../decorators";
+
+export class ProtonLiveSyncService implements IProtonLiveSyncService {
+	private excludedProjectDirsAndFiles = ["app_resources", "plugins", ".*.tmp", ".ab"];
+
+	private get $liveSyncServiceBase(): ILiveSyncServiceBase {
+		return this.$injector.resolve("liveSyncServiceBase");
+	}
+
+	constructor(private $devicesService: Mobile.IDevicesService,
+		private $errors: IErrors,
+		private $mobileHelper: Mobile.IMobileHelper,
+		private $options: ICommonOptions,
+		private $fs: IFileSystem,
+		private $injector: IInjector,
+		private $project: any,
+		private $logger: ILogger) { }
+
+	@exportedPromise("liveSyncService")
+	public livesync(deviceIdentifiers: IDeviceLiveSyncInfo[], projectDir: string, filePaths?: string[]): IFuture<void> {
+		return (() => {
+			this.$project.projectDir = projectDir;
+			let deviceInfos: any = _.map(deviceIdentifiers, d => {
+				let matchingDevice = _.find(this.$devicesService.getDeviceInstances(), device => device.deviceInfo.identifier === d.deviceIdentifier);
+				return _.extend(d, { device: matchingDevice });
+			});
+
+			let errors: Error[] = [];
+			// Group devices by platform. After that group each group by types of sync.
+			// Aggregate the errors and throw a single error in case any operation fails.
+			let groupedByPlatform = _.groupBy(deviceInfos, (d: any) => d.device.deviceInfo.platform.toLowerCase());
+			_.each(groupedByPlatform, (platformGroup: any[], platform: string) => {
+				let syncToApp = platformGroup.filter(plGr => !!plGr.syncToApp);
+				try {
+					this.liveSyncToApp(platform, syncToApp.map((s: any) => s.device), filePaths).wait();
+				} catch (err) {
+					this.$logger.trace(`Error while trying to livesync applications to devices: ${syncToApp.join(", ")}.`, err.message);
+					errors.push(err);
+				}
+
+				let syncToCompanion =  platformGroup.filter(plGr => !!plGr.syncToCompanion);
+				try {
+					this.liveSyncToCompanion(platform, syncToCompanion.map((s: any) => s.device), filePaths).wait();
+				} catch (err) {
+					this.$logger.trace(`Error while trying to livesync to companion app to devices: ${syncToCompanion.join(", ")}.`, err.message);
+					errors.push(err);
+				}
+			});
+
+			if(errors && errors.length) {
+				throw new Error(`Unable to livesync to specified devices. Errors are: ${errors.map(err => err.message)}`);
```

Comment:
It is not possible to implement such logic as you can throw error only with string message. My idea was to construct objects, for example:
```TypeScript
interface IDeviceLiveSyncError {
	deviceIdentifier: string;
	error: Error;
}
```
collect the errors and throw new Error with message: `JSON.stringify(deviceLiveSyncErrors)`
Unfortunately I'm not able to try-catch the errors per device at the moment, so I've decided to concatenate all error messages and implement the mentioned logic in a separate PR